### PR TITLE
Add matchers to LabelValues() call

### PR DESCRIPTION
**Pull request by replay**
This change improves the performance of the label-values lookup when matchers are specified ([API documented here](https://prometheus.io/docs/prometheus/latest/querying/api/#querying-label-values)). It does so by passing the matchers through to the index readers, which can implement the same functionality more efficiently.

This PR replaces (and is based on) https://github.com/prometheus/prometheus/pull/8361

I tested the performance difference by adding `1M` test metrics to a Prometheus instance. 
All metrics have a label `ninetyTen`, on `900k`/`90%` of them this label has the value `value1`, on the other `100k`/`10%` it has the value `value2`. 
All metrics also have a label `tenTen`, on the first `100k`/`10%` this label has the value `value0`, on the next `100k`/`10%` it has the value `value1`, etc such that each label value from `value0`-`value9` matches `100k`/`10%` of the metrics.
Then I query the label values of `tenTen` with the matcher `ninetyTen="value1"`, that matcher will match `900k`/`90%` of the metrics and the expected result set are the `tenTen` values `value0` - `value8`. I ran that query `10` times on `master` and `10` time on this branch.

```
time curl -g 'http://localhost:9090/api/v1/label/tenTen/values?match[]={ninetyTen="value1"}'
```

On `master` the latency range was `3.1s` - `3.3s`, on this branch the latency range was `1.5s` - `1.6s`. 
Note that this change might only make a relatively small difference for Prometheus itself, but it makes a big difference for the Cortex project once it also implements the `match` parameter on the label value API call, because without this change the Queriers would have to run a `.Select()` when matchers are specified and extract the label values from the returned metrics, these metrics would first get transmitted via the network and then they'd get sorted and de-duplicated, when a large number of metrics matches the matchers this becomes very expensive/slow. With this change the Queriers won't need to fetch & sort & de-duplicate all the metrics, because they can directly call the `.LabelValues()` method of the `Querier` interface and pass the matchers parameter.
I believe that the Thanos project could benefit from this in a similar way.

### CODE REVIEW

**Comment by replay** [comment]
This is related to https://github.com/cortexproject/cortex/issues/3658

**Comment by roidelapluie** [comment]
Hello, thanks for this pull requests.


Do we have benchmarks covering this ?

**Comment by replay** [comment]
> Do we have benchmarks covering this ?

Thx for your comment. 

A benchmark to effectively compare the performance of `master` against this branch would have to be in the `api` package because the `master` implementation of the matchers parameter is in the API, this benchmark would also need to generate a block and then instantiate a Querier for it.

Alternatively, I could just add a benchmark for the `LabelValues()` methods to show what their performance looks like. This would be relatively simple to do, but it would not be useful to compare the performance against `master`. 

Which of these two options do you think would be preferable? Would you consider it critical to be able to compare the performance against `master`?

**Comment by roidelapluie** [comment]
Wait, no, it is not needed. When I did look at first sight, it looked like this was a cortex only improvement. It turns out it is not, it is a different feature.

What prevents you to add the same parameter in the Prometheus API ?

**Comment by brian-brazil** [comment]
You could still add benchmarks for the new code so we can know if it regresses in future, and for the sake of this PR I think it would suffice to hack together something locally to give a rough idea of the improvement (even just using curl by hand) - no need to check it in or anything.

File: tsdb/head.go
Comment (roidelapluie) [inline tsdb/head.go:1638]: I think that should still be deferred ?

Comment (replay) [reply tsdb/head.go:1638]: the reason why i changed it non-deferred is because that way we can release the lock earlier on `:1666` if matchers have been specified. 
there is a trade-off between the safety of using deferred unlocks (because no-one can forget to unlock) and reducing lock congestion by unlocking early when it is possible.
not sure which side of this trade-off would be preferable, since the TSDB is a pretty integral and optimized part of Prometheus in general I chose the latter, but I'm happy to change this back if you disagree.

**Comment by replay** [comment]
> You could still add benchmarks for the new code so we can know if it regresses in future

I added benchmarks for the `LabelValues()` methods on the `blockIndexReader` and the `headIndexReader`.
The benchmarks are doing the `LabelValues()` lookup with a set of matchers matching `900k` metrics.

`blockIndexReader`
```
replay@nb-ubuntu:~/go/src/github.com/prometheus/prometheus$ go test  -run=^$ -bench '^(BenchmarkLabelValuesWithMatchers)$' 'github.com/prometheus/prometheus/tsdb'
goos: linux
goarch: amd64
pkg: github.com/prometheus/prometheus/tsdb
BenchmarkLabelValuesWithMatchers-8          2    701878734 ns/op        20841696 B/op    1800062 allocs/op
PASS
ok      github.com/prometheus/prometheus/tsdb   32.349s
```
`headIndexReader`
```
replay@nb-ubuntu:~/go/src/github.com/prometheus/prometheus$ go test  -run=^$ -bench '^(BenchmarkHeadLabelValuesWithMatchers)$' 'github.com/prometheus/prometheus/tsdb'
goos: linux
goarch: amd64
pkg: github.com/prometheus/prometheus/tsdb
BenchmarkHeadLabelValuesWithMatchers-8                 3         337562575 ns/op            3178 B/op         38 allocs/op
PASS
ok      github.com/prometheus/prometheus/tsdb   12.972s
```

I'll also come up with a curl-based benchmark to compare `master` against this branch and I'll post the details here.

**Comment by replay** [comment]
**Curl-based benchmark**

I generated a `metrics` file using the following Python script. It generates `1'000'000` metrics, out of which the first `900'000` have the label `ninetyTen="value1"`, the last `100'000` have `ninetyTen="value2"`.  
In addition each metric has a label `tenTen`, in the first `10%` of metrics this label has the value `value0`, in the next `10%` it has the value `value1`, etc up to `value9`.

```
#!/usr/bin/env python3

maxValue = 1000000
tenth = maxValue/10

for uniqueId in range(maxValue):
    labels = {
        'uniqueId': 'unique{uniqueId}'.format(uniqueId=uniqueId),
        'ninetyTen': 'value1' if uniqueId/tenth<9 else 'value2',
        'tenTen': 'value{bracket}'.format(bracket=int(uniqueId/tenth)),
    }
    print('test_metrics{' + ','.join(['{key}="{value}"'.format(key=key, value=value) for key, value in labels.items()]) + '} 123')
```

Then I made a Prometheus instance scrape that resulting `metrics` file and I left it running until it had generated `10` blocks with these `1'000'000` metrics.
Then I queried each of the two versions of Prometheus (`master` and this branch) `10` times with the following query, which looks up the label values of the label `tenTen` among all the metrics where `ninetyTen="value1"`. The expected output is `value0`-`value8` because the matchers will match the first `900'000` metrics but not the last `100'000`:

```
curl -g 'http://localhost:9090/api/v1/label/tenTen/values?match[]={ninetyTen="value1"}'
```

**master**
```
$ for i in {0..9}; do (time curl -g 'http://localhost:9090/api/v1/label/tenTen/values?match[]={ninetyTen="value1"}') 2>&1 | grep -E '^real'; done
real    0m28.915s
real    0m27.260s
real    0m27.599s
real    0m28.259s
real    0m28.379s
real    0m27.413s
real    0m27.997s
real    0m27.612s
real    0m27.616s
real    0m28.042s
```

**this branch**
```
$ for i in {0..9}; do (time curl -g 'http://localhost:9090/api/v1/label/tenTen/values?match[]={ninetyTen="value1"}') 2>&1 | grep -E '^real'; done
real    0m9.268s
real    0m8.501s
real    0m9.185s
real    0m8.571s
real    0m8.463s
real    0m8.544s
real    0m9.403s
real    0m8.534s
real    0m8.562s
real    0m9.245s
```

**Comment by replay** [comment]
@roidelapluie @brian-brazil 
Thanks for your comments above. This PR would be ready to review again, if you get a chance. I have posted the benchmarks which you asked for, please let me know if there's anything else I can do which would help to progress this.

**Review by cstyan — COMMENTED** [review]
This LGTM honestly, probably best if you have a quick look as tsdb maintainer @codesome

File: tsdb/block.go
Comment (roidelapluie) [inline tsdb/block.go:505]: Here and other comments

```suggestion
// LabelValueFor returns value of given label value of series referred to by ID.
```

File: tsdb/index/index.go
Comment (roidelapluie) [inline tsdb/index/index.go:1455]:
```suggestion
// Passing matchers is not supported yet.
```

should that be a TODO: ? otherwise, remove `yet`

File: tsdb/querier_test.go
Comment (roidelapluie) [inline tsdb/querier_test.go:1163]: I'd return early instead of else

File: tsdb/index/index.go
Comment (replay) [reply tsdb/index/index.go:1455]: yes, i'll make it a TODO

File: tsdb/block.go
Comment (roidelapluie) [reply tsdb/block.go]: note: there are similar comments like this :)

File: tsdb/querier_test.go
Comment (replay) [reply tsdb/querier_test.go:1163]: good idea, i'll do that

File: tsdb/block.go
Comment (replay) [reply tsdb/block.go]: sorry about that, i updated the remaining ones https://github.com/prometheus/prometheus/pull/8400/commits/ed646466c4de35d44ddbb266a3342241bc997a3f

**Comment by replay** [comment]
FYI I re-based onto the latest `master` and force pushed

File: tsdb/head.go
Comment (codesome) [inline tsdb/head.go:1662]: Based on what I commented about the `symMtx`, we can use the `LabelValueFor` here and maybe even share the code between head.go and block.go to do this.

File: tsdb/index/index.go
Comment (codesome) [inline tsdb/index/index.go:1513]: Same comment modifications as suggested somewhere else

File: tsdb/index/index.go
Comment (codesome) [inline tsdb/index/index.go:1731]: This should go inside `if ln == label {`

File: tsdb/index/index.go
Comment (codesome) [inline tsdb/index/index.go:1738]:
```suggestion
	return "", d.Err()
```

File: tsdb/index/index.go
Comment (codesome) [inline tsdb/index/index.go:1522]: Maybe do the `d.Get()` before this and then check the `d.Err()` error. We should do after `d.Get()` anyway.

File: storage/merge_test.go
Comment (replay) [reply storage/merge_test.go:984]: oh, yes it should, thanks!

Comment (replay) [reply storage/merge_test.go]: fixed in https://github.com/prometheus/prometheus/pull/8400/commits/4b9220d06f69915da6c992c770235cd1847da652

File: tsdb/block.go
Comment (replay) [reply tsdb/block.go]: Thx fixed in https://github.com/prometheus/prometheus/pull/8400/commits/c2f560728cbec5ea14c276fa2f1ca06c8232e079

File: tsdb/index/index.go
Comment (replay) [reply tsdb/index/index.go]: thx, fixed in https://github.com/prometheus/prometheus/pull/8400/commits/c2f560728cbec5ea14c276fa2f1ca06c8232e079

File: tsdb/head.go
Comment (replay) [reply tsdb/head.go]: thx, fixed in bacbf539f10b3a289e364b8e24f22938824d8d22

File: tsdb/index/index.go
Comment (replay) [reply tsdb/index/index.go]: thx, fixed cdbc0027f957a0c4329db915c998f25371d920dd

File: tsdb/index/index.go
Comment (replay) [reply tsdb/index/index.go:1523]: is this what you had in mind? d2191c5a5e67d373e0f972a173e6fcb5cfcab578

File: tsdb/head.go
Comment (replay) [reply tsdb/head.go]: Ah really, sorry I guess I'm confused about what exactly is protected by this lock. In this case this should be fine? 4110d3677f8fb310e39b5883334c6e562a215e93

The function `h.head.postings.LabelValues()` seems to have its own lock-handling, so what is `h.head.symMtx` protecting which is read in this line? Is it just to read the pointer value of `h.head.postings`?

File: tsdb/head.go
Comment (replay) [reply tsdb/head.go]: you're right, thx. fixed in 4940e15db408e7516e44b28631e972d3bbfc9aed

File: tsdb/head.go
Comment (replay) [reply tsdb/head.go]: That's a very nice idea, how about that: d91d7f2d93eaa25a085627d618722073025b9b50

I didn't really know where to put this `labelValuesWithMatchers()` function because its supposed to be generic and not specific to either the head or the block, so i added a `util.go` file... Do you have a better idea?

File: tsdb/block.go
Comment (replay) [reply tsdb/block.go:91]: Good point, the two implementations are currently differing in their behavior in that case. I'll change it so an error gets returned whenever the series does not have the given label name.

Comment (replay) [reply tsdb/block.go:91]: Fixed here: 05f431417d355a0af8c5d6c479d82857ef64fd1a

File: tsdb/index/index.go
Comment (codesome) [reply tsdb/index/index.go:1523]: Yes

File: tsdb/head.go
Comment (codesome) [reply tsdb/head.go]: Since `PostingsForMatchers` is in querier.go and as this is related to queries, lets get rid of util.go and put it in querier.go?

File: storage/merge_test.go
Comment (codesome) [inline storage/merge_test.go:983]:
```suggestion
				require.Equal(t, tcase.expectedWarnings[3], w)
```

File: tsdb/block.go
Comment (codesome) [reply tsdb/block.go:91]: Let's return a named error when we have no series or do not find any label so the we need not hard fail on them. We have `storage.ErrNotFound` which we can use for both (let us also comment here that it returns `storage.ErrNotFound` when series or the label is not found).

Note that you need to now look for that error in `labelValuesWithMatchers` and not hard fail (Head compaction might have some races where the series is garbage collected, so lets be safe)

File: web/api/v1/api.go
Comment (codesome) [inline web/api/v1/api.go:625]: I think we still need this :) looks like the tests are not covering this, could you add a test case covering this?

**Review by codesome — COMMENTED** [review]
Almost there!

File: tsdb/head.go
Comment (codesome) [reply tsdb/head.go]: symMtx is actually protecting the symbols in Head but also protecting the race between adding/deleting in h.postings and updating the symbols map in Head since they need to be in sync, hence you see h.postings being modified under that lock too.

File: tsdb/head.go
Comment (replay) [reply tsdb/head.go]: good idea, done https://github.com/prometheus/prometheus/pull/8400/commits/3c3b76a02a95f6aa64d1d100f28793fa63ad9292

File: web/api/v1/api.go
Comment (replay) [reply web/api/v1/api.go:625]: I think we can definitely remove the `labelValuesSet` map, as this isn't even read anymore: 2e10b005f6bb1f77387f232227f5146438863f2f

There are actually unit tests for this function, which are also using matchers:
https://github.com/prometheus/prometheus/blob/06e53b7ecfca1350e73a6e75f789313ae01139d1/web/api/v1/api_test.go#L721-L775

But currently the results returned from `LabelValues()` in the unit tests are always sorted because the tests are using the `blockBaseQuerier`, which in its `LabelValues()` calls `q.index.SortedLabelValues()`: https://github.com/prometheus/prometheus/blob/3c3b76a02a95f6aa64d1d100f28793fa63ad9292/tsdb/querier.go#L87

Actually I don't know why the `blockBaseQuerier.LabelValues()` method would call `q.index.SortedLabelValues()` and I think it could just call `q.index.LabelValues()` instead, but I didn't add that so I'm hesitant to just change it. If I want to create a test which checks whether `api.labelValues()` sorts the list of values I'll have to make it use some alternative querier because as long the test is using the `blockBaseQuerier` the list of values would be sorted anyway.

Comment (replay) [reply web/api/v1/api.go]: good catch, thx.

I added a unit test and a fix in `api.go`, without the fix in `api.go` the unit test was failing: 3a3290f8fccba86ee1b93144c033303b6d53ab2e

File: tsdb/block.go
Comment (replay) [reply tsdb/block.go:93]: Is this what you had in mind: 9805c9ded

**Review by codesome — APPROVED** [review]
LGTM, Thanks! I will merge on green.

**Pull request merged**
